### DISCUSSION TITLE
By using parse_key_value the key-values are not getting into the output

### DISCUSSION THREAD

**Discussion by uni-pooja-laad** [vectordotdev/vector#21824] [Q&A]
Hi team,
in contineousion of https://discord.com/channels/742820443487993987/1288128207111065650/1295892531246268540
I am using remap with parse_key_value. The vector configuration file attached. The log coming from datadog_agent attached in the file orch_swepper_logs.yaml. 
In Datadog it is appearing without tags attached screenshot. (filename: after.png)
Before applying this remap, the tags were getting properly, that Screenshot too attached(filename: before.png)
Why this is happening? I just used it to parse ddtags key and fetch kube_namespace from it to match with the condition.
![after](https://github.com/user-attachments/assets/babff63f-8d5e-4cde-875a-ebf2365a4f8f)
![before](https://github.com/user-attachments/assets/3ceac874-e999-4d0d-a4b7-000b73724eaa)
[orch_swepper_logs.json](https://github.com/user-attachments/files/17801979/orch_swepper_logs.json)
[vector-config-ai.txt](https://github.com/user-attachments/files/17802016/vector-config-ai.txt) 
Attaching vector-config-ai as txt as json format not supported here.

**Comment by pront** [comment]
Hi @uni-pooja-laad,

A few things that might help here. The logs intake API expects something [like this](https://docs.datadoghq.com/api/latest/logs/?code-lang=python), specifically this line:
```
ddtags="env:staging,version:5.1",
```

Also, see this [playground example](https://playground.vrl.dev/?state=eyJwcm9ncmFtIjoiLmRkdGFncyA9IHBhcnNlX2tleV92YWx1ZSEoLmRkdGFncywgZmllbGRfZGVsaW1pdGVyOiBcIixcIiwga2V5X3ZhbHVlX2RlbGltaXRlcjogXCI6XCIpXG4uaXNfYWkgPSAuZGR0YWdzLmt1YmVfbmFtZXNwYWNlID09IFwiYWlcIiIsImV2ZW50Ijp7ImNsdXN0ZXJfbmFtZSI6InVuaXBob3JlLXN0YWdpbmctdXMtZWtzLXBsYXRmb3JtIiwiZGRzb3VyY2UiOiJ1em5nLW9yY2hlc3RyYXRvci1zd2VlcGVyIiwiZGR0YWdzIjoiZmlsZW5hbWU6MC5sb2csZGlybmFtZTovdmFyL2xvZy9wb2RzL3VhbmFseXplLW5nX3V6bmctb3JjaGVzdHJhdG9yLXN3ZWVwZXItN2NiYjY2ZDc0NS1qa2t2aF82MGJmMDNhNy1kNGEyLTQ2NDktYTFkNi03NTU4ZmM0YTc2NzcvdXpuZy1vcmNoZXN0cmF0b3Itc3dlZXBlcixrdWJlX2NvbnRhaW5lcl9uYW1lOnV6bmctb3JjaGVzdHJhdG9yLXN3ZWVwZXIsaW1hZ2VfbmFtZTpyZWdpc3RyeS51bmlwaG9yZS5jb20vdW5pcGhvcmUvdXpuZy1vcmNoZXN0cmF0b3Itc3dlZXBlcixrdWJlX2FwcF9pbnN0YW5jZTp1em5nLW9yY2hlc3RyYXRvci1zd2VlcGVyLGt1YmVfb3duZXJyZWZfa2luZDpyZXBsaWNhc2V0LGt1YmVfcmVwbGljYV9zZXQ6dXpuZy1vcmNoZXN0cmF0b3Itc3dlZXBlci03Y2JiNjZkNzQ1LGt1YmVfbmFtZXNwYWNlOnVhbmFseXplLW5nLGt1YmVfcW9zOkJ1cnN0YWJsZSxrdWJlX2FwcF9tYW5hZ2VkX2J5OkhlbG0sa3ViZV9hcHBfbmFtZTp1em5nLW9yY2hlc3RyYXRvci1zd2VlcGVyLHBvZF9waGFzZTpydW5uaW5nLHNlcnZpY2U6dXpuZy1vcmNoZXN0cmF0b3Itc3dlZXBlcixpbWFnZV9pZDpyZWdpc3RyeS51bmlwaG9yZS5jb20vdW5pcGhvcmUvdXpuZy1vcmNoZXN0cmF0b3Itc3dlZXBlckBzaGEyNTY6NTc1MjU1NWQ4MGRkZTdiZDFjMzRkMmE4NjBiYWQ5ODU1NjU2NmRmYzg3MWRjMDVjMzI0ZTFkMDA4YzBkZWQ0YixlbnY6c3RhZ2luZyx2ZXJzaW9uOnYtZDQ2YjI5N2MxZSxrdWJlX2RlcGxveW1lbnQ6dXpuZy1vcmNoZXN0cmF0b3Itc3dlZXBlcixzaG9ydF9pbWFnZTp1em5nLW9yY2hlc3RyYXRvci1zd2VlcGVyLGltYWdlX3RhZzp2LWQ0NmIyOTdjMWUscG9kX25hbWU6dXpuZy1vcmNoZXN0cmF0b3Itc3dlZXBlci03Y2JiNjZkNzQ1LWpra3ZoLGt1YmVfb3duZXJyZWZfbmFtZTp1em5nLW9yY2hlc3RyYXRvci1zd2VlcGVyLTdjYmI2NmQ3NDUsY29udGFpbmVyX2lkOmNmNTJiMTVkYjJmMjFiZTFkN2UzMzdlNGRhYjFkNWZiNTZlNDhlZWM5NzI2Yzk0OTBhNWIwYjRhZWIxOTIwMmYsZGlzcGxheV9jb250YWluZXJfbmFtZTp1em5nLW9yY2hlc3RyYXRvci1zd2VlcGVyX3V6bmctb3JjaGVzdHJhdG9yLXN3ZWVwZXItN2NiYjY2ZDc0NS1qa2t2aCxjb250YWluZXJfbmFtZTp1em5nLW9yY2hlc3RyYXRvci1zd2VlcGVyIiwiZW52Ijoic3RhZ2luZyIsImhvc3RuYW1lIjoiaS0wNTdmYmM1MjM2ZWU4ZDRkNCIsIm1lc3NhZ2UiOiJ7XCJ0aW1lc3RhbXBcIjpcIjIwMjQtMTEtMThUMTQ6MzU6MzQuODA4WlwiLFwibWVzc2FnZVwiOlwiY3JvbiB1cGRhdGUgcHJvY2VzcyBzdGF0dXMgZW5kZWRcIixcImxvZ2dlclwiOlwiY29tLnVuaXBob3JlLm9yY2hlcy5zd2VlcGVyLmpvYi5VcGRhdGVQcm9jZXNzaW5nU3RhdHVzSm9iXCIsXCJ0aHJlYWRcIjpcInNjaGVkdWxpbmctMVwiLFwibGV2ZWxcIjpcIklORk9cIixcImNvbnZlcnNhdGlvbklkXCI6XCI0OGU5MTEwNy0wOTRiLTQzOWItYTFiOS03OTJjZjIwOTc3ZDJcIixcImRkLnRyYWNlX2lkXCI6XCI2NjI4OTg1NDc2ODE5NzcyNDQ5XCIsXCJkZC5zZXJ2aWNlXCI6XCJ1em5nLW9yY2hlc3RyYXRvci1zd2VlcGVyXCIsXCJkZC5lbnZcIjpcInN0YWdpbmdcIixcImRkLnZlcnNpb25cIjpcInYtZDQ2YjI5N2MxZVwiLFwiZGQuc3Bhbl9pZFwiOlwiODgxODYyNzI3ODAyMzUxNTk1NFwifSIsInJlZ2lvbiI6InVzIiwic2FtcGxlX3JhdGUiOiIxIiwic2VydmljZSI6InV6bmctb3JjaGVzdHJhdG9yLXN3ZWVwZXIiLCJzb3VyY2VfdHlwZSI6ImRhdGFkb2dfYWdlbnQiLCJzdGF0dXMiOiJpbmZvIiwidGltZXN0YW1wIjoiMjAyNC0xMS0xOFQxNDozNTozNy4wNDAwODQyNzNaIiwidmVjdG9yX3RhZyI6InllcyJ9LCJpc19qc29ubCI6ZmFsc2UsImVycm9yIjpudWxsfQ%3D%3D):

```
# convert to an object from a string
.ddtags = parse_key_value!(.ddtags, field_delimiter: ",", key_value_delimiter: ":")
```
---

The JSON sample you provided correctly specifies the `.ddtags` as a command separated list of key/value pairs. Then, your remap will transform them to:

```
"ddtags": {
  "container_id": "cf52b15db2f21be1d7e337e4dab1d5fb56e48eec9726c9490a5b0b4aeb19202f",
  "container_name": "uzng-orchestrator-sweeper",
  "dirname": "/var/log/pods/uanalyze-ng_uzng-orchestrator-sweeper-7cbb66d745-jkkvh_60bf03a7-d4a2-4649-a1d6-7558fc4a7677/uzng-orchestrator-sweeper",
  "display_container_name": "uzng-orchestrator-sweeper_uzng-orchestrator-sweeper-7cbb66d745-jkkvh",
  "env": "staging",
  "filename": "0.log",
  "image_id": "registry.uniphore.com/uniphore/uzng-orchestrator-sweeper@sha256:5752555d80dde7bd1c34d2a860bad98556566dfc871dc05c324e1d008c0ded4b",
  "image_name": "registry.uniphore.com/uniphore/uzng-orchestrator-sweeper",
  "image_tag": "v-d46b297c1e",
  "kube_app_instance": "uzng-orchestrator-sweeper",
  "kube_app_managed_by": "Helm",
  "kube_app_name": "uzng-orchestrator-sweeper",
  "kube_container_name": "uzng-orchestrator-sweeper",
  "kube_deployment": "uzng-orchestrator-sweeper",
  "kube_namespace": "uanalyze-ng",
  "kube_ownerref_kind": "replicaset",
  "kube_ownerref_name": "uzng-orchestrator-sweeper-7cbb66d745",
  "kube_qos": "Burstable",
  "kube_replica_set": "uzng-orchestrator-sweeper-7cbb66d745",
  "pod_name": "uzng-orchestrator-sweeper-7cbb66d745-jkkvh",
  "pod_phase": "running",
  "service": "uzng-orchestrator-sweeper",
  "short_image": "uzng-orchestrator-sweeper",
  "version": "v-d46b297c1e"
}
```

The above allows you to access the individual tags by keys e.g. `.ddtags.kube_app_name`.

~Finally, the datadog logs sink will internally take this object and convert it back to a single string~. Are you modifying the above object in any other way after the `parse_ai_ns` transform?

**Comment by uni-pooja-laad** [comment]
It worked by assigning parsed values to new variable and then using it: 

      parse_ai_ns:
        type: remap
        inputs: [filter_ai_ns]
        source: |-
          .datadogtags = parse_key_value!(.ddtags, field_delimiter: ",", key_value_delimiter: ":")
          .timestamp = now()
           .vector_tag = "yes"
      get_ai_ns:
        inputs: [parse_ai_ns]
        type: filter
        condition: 
          type: "vrl"
          source: .datadogtags.kube_namespace=="ai"

**Reply by pront** [reply]
> It worked by assigning parsed values to new variable

This works because you left the original `ddtags` as a string with the expected format. Also, I looked at the implementation here:
https://github.com/vectordotdev/vector/blob/master/src/sinks/datadog/logs/sink.rs#L112-L129 

and it turns out the sink expects the tags to be a string or an array. So the following object won't be converted to the correct format:
```
"ddtags": {
  "container_id": "cf52b15db2f21be1d7e337e4dab1d5fb56e48eec9726c9490a5b0b4aeb19202f",
  "container_name": "uzng-orchestrator-sweeper",
  ...
}
```

Arguably we could make the sink smarter but that would be a separate enhancement request.

If you had an array in the event e.g. `"ddtags": [ "container_id:foo", "container_name:bar", ... ]` then it would have been converted to the correct format inside the sink. 

Hopes this answer covers everything.
